backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/ModularContractService.java

```diff
@@ -399,4 +406,73 @@ public void cancelContractsOnOrderVoidIfNeededAndAllowed(@NonNull final OrderId
 													.build());
 		}
 	}
+
+	public void closeContractsOnOrderCloseIfNeededAndAllowed(@NonNull final I_C_Order orderRecord)
+	{
+		final List<de.metas.interfaces.I_C_OrderLine> orderLines = orderBL.retrieveOrderLines(orderRecord);
+		final Set<FlatrateTermId> modularContractIds = orderLines.stream()
+				.filter(this::isModularContractLine)
+				.map(I_C_OrderLine::getC_Order_ID)
+				.map(OrderId::ofRepoId)
+				.map(flatrateBL::getByOrderId)
+				.flatMap(List::stream)
+				.map(I_C_Flatrate_Term::getC_Flatrate_Term_ID)
+				.map(FlatrateTermId::ofRepoId)
+				.collect(Collectors.toSet());
+		final boolean isModularOrder = !modularContractIds.isEmpty();
+
+		if (!isModularOrder)
+		{
+			return;
+		}
+
+		final boolean hasBillableUnprocessedLogs = modularContractLogService.anyMatch(
+				ModularContractLogQuery.builder()
+						.flatrateTermIds(modularContractIds)
+						.billable(true)
+						.processed(false)
+						.build()
+		);
+
+		if(hasBillableUnprocessedLogs)
+		{
+			throw new AdempiereException("Not all billable logs are processed"); // TODO
+		}
+
+		if(!orderRecord.isSOTrx())
+		{
+			final Set<OrderId> relatedSalesOrdersIds = queryBL.createQueryBuilder(I_C_OrderLine.class)
```

Comment:
pls don't call `queryBL.createQueryBuilder(I_C_OrderLine.class)`.

i suggest to introduce a method in order DAO & BL